[ENH] Update ME-EPI workflow to create optimal combination

emdupre opened the pull request:
<!--
Text in these brackets are comments, and won't be visible when you submit your pull request.
If this is your first contribution, please take the time to read these, in particular the comment
beginning "Welcome, new contributors!".
-->

## Changes proposed in this pull request

- Removes in-package implementation of T2* map, wraps tedana in a nipype style interface

- IMPORTANT ! Integrates optimal combination into fmriprep workflow as a default. Previously, users would supply multi-echo files and receive all processed, individual echos back. Now, users will receive a single, optimally combined time series when using multi-echo files. This is in response to queries like [this one on NeuroStars](https://neurostars.org/t/fmriprep-combining-multi-echo-data-error-with-unwarping/2310/3), and several offline requests. Crucially, however, this effectively removes the ability for users to run fmriprep and _then_ tedana, as tedana denoising will now need to be _integrated_ into fmriprep entirely, probably by subbing out the AROMA connection. I would, though, like to have this working before opening another PR for adding in denoising.

<!--
Please describe here the main features / changes proposed for review and integration in fMRIPrep
If this PR addresses some existing problem, please use GitHub's citing tools 
(eg. ref #, closes # or fixes #).
If there is not an existing issue open describing the problem, please consider opening a new
issue first and then link it from here (so the *fMRIPrep* community has a better understanding
of ongoing development efforts and possible overlaps between contributions).
-->


## Documentation that should be reviewed
<!--
Please summarize here the main changes to the documentation that the reviewers should be aware of.
-->
Update docstring for `T2SMap`, also updated workflow description for `bold_t2s_wf`. In both cases, removed graphs (as I considered these unhelpful, since so many of the steps are wrapped within `tedana`). 



<!--
Welcome, new contributors!

We ask you to read through the Contributing Guide:
https://github.com/poldracklab/fmriprep/blob/master/CONTRIBUTING.md

These are guidelines intended to make communication easier by describing a consistent process, but
don't worry if you don't get it everything exactly "right" on the first try.

To boil it down, here are some highlights:

1) Consider starting a conversation in the issues list before submitting a pull request. The discussion might save you a
   lot of time coding.
2) Please use descriptive prefixes in your pull request title, such as "ENH:" for an enhancement or "FIX:" for a bug fix.
   (See the Contributing guide for the full set.) And consider adding a "WIP" tag for works-in-progress.
3) Any code you submit will be licensed under the same terms (BSD 3-Clause) as the rest of fMRIPrep.
4) We invite every contributor to add themselves to the `.zenodo.json` file
   (https://github.com/poldracklab/fmriprep/blob/master/.zenodo.json), which will result in your being listed as an author
   at the next release. Please add yourself as the next-to-last entry, just above Russ.

A pull request is a conversation. We may ask you to make some changes before accepting your PR,
and likewise, you should feel free to ask us any questions you have.

-->

emdupre commented on fmriprep/workflows/bold/base.py:
```diff
@@ -561,28 +560,31 @@ def init_func_preproc_wf(bold_file, ignore, freesurfer,
+    workflow.connect([
+        # Connect bold_bold_trans_wf -- either original or T2S shadow
+        (inputnode, bold_bold_trans_wf, [
+            ((if multiecho 'ref_file' else 'bold_file'), 'inputnode.name_source')]),
```
I'm a little stuck here. I seem to remember seeing something like this at one point (maybe @effigies or @oesteban had done something similar ?), but I can't figure out a nipype appropriate way to do this. Any pointers would be appreciated !

effigies replied:
This is a ternary operator with a minor syntax error. Try `'ref_file' if multiecho else 'bold_file'`.

emdupre replied:
That was it ! Sorry, I had it set up like an if-statement. Thanks !

emdupre commented:
I've tried to be very explicit in this code (so far) about disconnecting and re-connecting workflows if multi-echo data is provided. I'd like to think there's a cleaner way to do this, but it seems that it would involve shadowing the original `bold_trans_trans_wf` and I'm not sure that's best practice. Any thoughts one way or another would be appreciated !

Currently I'm running up against [an unhelpful error message](https://github.com/nipy/nipype/issues/2245#issue-267309769) for the multi-echo test case, but I'll pick back up again with this tomorrow.

effigies commented on fmriprep/interfaces/multiecho.py:
```diff
@@ -78,320 +74,44 @@ def _run_interface(self, runtime):
-        neg_tes = [-1000 * te for te in tes[:echo]]
+        (self._results['t2sv'], self._results['s0v'],
+         self._results['t2svG'], self._results['s0vG'],
+         self._results['ts_OC']) = tedana.t2smap_workflow(self.inputs.in_files, echo_times)
```
I'm wondering if the issue is here. [`t2smap_workflow`](https://github.com/ME-ICA/tedana/blob/e7f04ff299412befbf309ceece17cd6e4889038e/tedana/workflows/t2smap.py#L78-L194) does not seem to return anything.

I'm not sure why this would result in a failure like the issue you linked to, but sometimes things cascade weirdly.

emdupre replied:
You're totally right -- it writes out the files rather than returning a value. I'll go back to some of the other nipype interfaces and check to see how they're capturing outputs.

effigies commented on fmriprep/interfaces/multiecho.py:
```diff
@@ -78,320 +74,44 @@ def _run_interface(self, runtime):
 class T2SMapInputSpec(BaseInterfaceInputSpec):
-    in_files = InputMultiPath(File(exists=True), mandatory=True,
+    in_files = InputMultiPath(File(exists=True), mandatory=True, minlen=2,
                               desc='multi-echo BOLD EPIs')
```
If there's a minimum length of 2, then this should just be a `traits.List`, not an `InputMultiPath`.

emdupre replied:
It looks like `InputMultiPath` this use in the [`FirstEcho` interface](https://github.com/poldracklab/fmriprep/blob/f5b502c329daef0a1d0766c1553d08cb65115af9/fmriprep/interfaces/multiecho.py#L34) as well -- should I update both ?

effigies replied:
Yes.

effigies commented on fmriprep/interfaces/multiecho.py:
```diff
@@ -78,320 +74,44 @@ def _run_interface(self, runtime):
+    s0v = File(exists=True, desc='limited s0 map')
+    t2svG = File(exists=True, desc='adaptive T2* map')
+    s0vG = File(exists=True, desc='adaptive s0 map')
+    ts_OC = File(exists=True, desc='optimally combined ME-EPI time series')
```
I would prefer more descriptive output names, such as `t2star_map`, `s0_map`, etc.

emdupre replied:
Yes, I can definitely do that ! Sorry, these were taken from tedana itself.

emdupre commented:
Ok, I've updated the interface to address @effigies review comments (thank you !), but I'm still receiving the same error. I'm still mostly sure it's related to the `JoinNode`, or at least an `IdentityInterface`. Will update as I work through this !

emdupre commented:
Just as an update, .@effigies was able to solve my error ! From the slack channel:
> My guess is what’s happening is that when you disconnect `bold_bold_trans_wf`, the nodes still exist in the workflow but aren’t receiving inputs. If that’s the case, probably what’s needed is to conditionally connect, rather than connect and then conditionally disconnect.
Also if that’s true, it’s a really weird way for the error to express, and probably worth seeing if we can detect situations like that better in nipype.

This was exactly the issue -- I've rearranged the `bold_t2s_wf` so that now `bold_bold_trans_wf` is called outside of the T2* workflow and unconnected nodes are no longer present.

I'm now working on (what I think is an unrelated issue):
```
Standard error:
Traceback (most recent call last):
  File "/usr/local/miniconda/bin/t2smap", line 11, in <module>
    sys.exit(_main())
  File "/usr/local/miniconda/lib/python3.6/site-packages/tedana/workflows/t2smap.py", line 200, in _main
    if options.debug and not options.quiet:
AttributeError: 'Namespace' object has no attribute 'debug'
```

Will report back.

effigies commented:
Via https://mail.nmr.mgh.harvard.edu/pipermail/freesurfer/2018-September/058436.html

> Unfortunately, the disk that hosts the ftp data has died, so resolving this might take a bit longer than expected. The new disk should be repopulated within the next few days.

emdupre commented:
OK, I know this isn't passing CI because of the Freesurfer FTP issue, but I think this is actually ready for review !

It's passing locally (see included screenshot) and now we're down to one output file for all provided echos; i.e., we're getting the optimal combination back, so users could call AROMA or whatever they'd like on it until we integrate tedana. Slice timing is also finally incorporated for MEEPI data, so I'm sure users will be happy to be past that error.

The T2* coreg flag is still performing worse than I would have expected, but better than before. I'd prefer to address that in a different PR, since this one is so big already ! But LMK what you think is easiest / best.

![screenshot_2018-09-05 screenshot](https://user-images.githubusercontent.com/15017191/45103149-378cf800-b0fd-11e8-89ff-445fc49534fe.png)

emdupre commented on fmriprep/workflows/bold/base.py:
```diff
@@ -539,48 +528,69 @@ def init_func_preproc_wf(bold_file, ignore, freesurfer,
+            workflow.disconnect([
+                (bold_reference_wf, bold_stc_wf,
+                    [('outputnode.bold_file', 'inputnode.bold_file')])])
+            meepi_echos = boldbuffer.clone(name='meepi_echos')
```
Is there a way to pass iterables to `LiterateWorkflow` directly ?

emdupre replied:
Pinging @effigies as the nipype guru :)

effigies replied:
Treat `LiterateWorkflow` the same as any other `Workflow`. Or is the question how to pass iterables to any workflow?

emdupre replied:
Yes, please.

emdupre replied:
Resolved in person -- the used approach is best for now.

romainVala commented:
Hi

I am very interested by this pull request, and so thanks for sharing your work
I came to fmriprep to preprocess the multiecho data in order to run the tedana denoising. So I am surprise by your choice

>     * however, this effectively removes the ability for users to run fmriprep and _then_ tedana, as tedana denoising will now need to be _integrated_ into fmriprep entirely, probably by subbing out the AROMA connection.

I understand that it will be easier to have all step include in fmriprep, but it is important to have some critical preprocessing step in order to be able to run different pipeline combination

I do not agree to have only the optimal echo recombination (by the way what is the optimal combination : the SNR weighted sum of the exp(-T2/TE) weighted sum ?). 
Why not just produce both (each preprocessed echo and the  optimally combined echo ) and later the denoise version. 
May be just a command line argument that will give you the choice to have all.

May be I am wrong, and I can understand the main purpose of fmriprep which is to produce the 'best preprocessed' finale time series, but I thing it is not incompatible with the possibilité to provide specific intermediary results.

Cheers

emdupre commented:
Hi @romainVala, thanks for sharing your thoughts ! This PR is not meant to be standalone; the plan is that we will *also* incorporate tedana denoising directly into fMRIPrep in a separate PR.

To make this a little clearer (and to prevent releasing a version with only partial multiecho support), @effigies has created a `multiecho` branch where I'll continue this work. The thought, then, would be that if you would like the individual echos returned, we'll add in an `--echo_id` flag (similar to the `--task_id` flag) where you can select individual echos for processing.

Let me know if this addresses your concerns !

effigies commented:
Alas, conflicts to merge...

emdupre commented:
Assuming I handled this merge conflict correctly, I'm going to add one more commit re: using #1278 and then I think this is ready for review !

effigies commented on fmriprep/workflows/bold/base.py:
```diff
@@ -533,19 +556,55 @@ def init_func_preproc_wf(bold_file, ignore, freesurfer,
+            (bold_t2s_wf, bold_reg_wf, [
+                ('outputnode.bold_ref', 'inputnode.ref_bold_brain'),
+                ('outputnode.bold_mask', 'inputnode.ref_bold_mask')]),
+        ])
```
Just a quick note, it's the now-missing `bold_reg_wf.inputnode.ref_bold_brain` that's causing the test failures. So you may want to grep through for other cases like this.

emdupre commented:
I used to call `bold_bold_trans_wf` _in_ `bold_t2s_wf`, since we need the individual echos to be motion corrected before creating the optimal combination. I moved it to just before `bold_t2s_wf` to get around the `TypeError("'_Undefined' object does not support indexing ...")` issue we were seeing earlier when I wasn't fully disconnecting the node.

As to why to not feed the inputs into `bold_bold_trans_wf`, it would effectively be trying to derive motion parameters on motion-corrected files, so I thought it would be better to avoid that extra interpolation. Otherwise we'll have to run it twice (once before and once after `bold_t2s_wf`). 

Does that make sense ? WDYT ?

effigies commented:
Ah, right. Sorry, I forgot about the whole motion correction thing. Let me read through again with that in mind.

emdupre commented on fmriprep/utils/misc.py:
```diff
@@ -24,6 +24,26 @@ def fix_multi_T1w_source_name(in_files):
+    import os
+    from nipype.utils.filemanip import filename_to_list
+    base, in_file = os.path.split(filename_to_list(in_files)[0])
+    keys_pre_echo = in_file.split("_")[:-2]
```
in the current spec, echo is guaranteed to be last ! happy to update if you'd like to future-proof, though

effigies replied:
Either way. If this works for all valid datasets, then I'm okay with it.

emdupre commented on fmriprep/workflows/bold/base.py:
```diff
@@ -438,20 +432,32 @@ def init_func_preproc_wf(bold_file, ignore, freesurfer,
+                (bold_reference_wf, boldbuffer, [
+                    ('outputnode.bold_file', 'bold_file')])])
+        else:  # for meepi, iterate over all meepi echos to boldbuffer
+            boldbuffer.iterables = ('bold_file', bold_file)
```
Sorry, I don't understand this suggestion-- I don't think I can have an `elif / else` block without an `if` statement ?

effigies replied:
Sorry for being unclear. I mean the current structure is:

```Python
if run_stc is True:
    # block A
else:
    if not multiecho:
        # block B
    else:
        # block C
```

This could become either:

```Python
if run_stc is True:
    # block A
elif not multiecho:
    # block B
else:
   # block C
```

or:

```Python
if run_stc is True:
    # block A
elif multiecho:
    # block C
else:
    # block B
```

emdupre replied:
I see, but there are really four options, no ? It's all combinations of `run_stc is True` and `multiecho is True`. So I could do something like:

```
if run_stc is True and multiecho is True:
    # block A
elif run_stc is False and multiecho is True:
    # block C
...
```

emdupre commented on fmriprep/workflows/bold/base.py:
```diff
@@ -471,6 +477,24 @@ def init_func_preproc_wf(bold_file, ignore, freesurfer,
+                                       mem_gb=mem_gb['resampled'],
+                                       omp_nthreads=omp_nthreads,
+                                       name='bold_t2smap_wf')
+
```
Moved !

emdupre commented:
The T2* map creation and the optimal combination are done in one step, so each echo needs to be SDC'ed separately. Right now I'm trying to think through the best way to combine the transforms, though, since I need the optimal combination to be in the same space as the T1w..

emdupre commented on fmriprep/workflows/bold/base.py:
```diff
@@ -451,13 +451,13 @@ def init_func_preproc_wf(bold_file, ignore, freesurfer,
-                    ('outputnode.bold_file', 'bold_file')])])
-        else:  # for meepi, iterate over all meepi echos to boldbuffer
-            boldbuffer.iterables = ('bold_file', bold_file)
+    elif (run_stc is False) and not multiecho:
```
You're right ! Sorry, fixed.

effigies commented:
> The T2* map creation and the optimal combination are done in one step, so each echo needs to be SDC'ed separately.

Okay, then let's keep things as they are, for now.

> Right now I'm trying to think through the best way to combine the transforms, though, since I need the optimal combination to be in the same space as the T1w.

Not knowing details of tedana, I apologize for any absurdities that follow:

Perhaps it could be written as a two-step process, where the T2* map is created as well as the transform (mixing matrix?) from the BOLD series, but the optimal combination is not generated. Then we could run SDC and coregistration on the T2*, and pass in the motion parameters, warp field, and affine transform (and possibly T1-template warp) to perform a single shot optimal-combination construction.

effigies commented:
If I'm reading that last correctly, you're just disabling SDC for multiecho? If so, then it probably makes sense to just do that at the bold workflow level, and let the existing logic take care of handling the transforms.

emdupre commented:
I don't think it's being totally disabled (if I'm understanding [the processing tree](https://fmriprep.readthedocs.io/en/stable//workflows-4.pdf) correctly) since SDC is still feeding into `bold_bold_trans_wf`, which is what I'm using to derive processed echos from which the optimal combination is generated. Does that match your understanding ?

effigies commented:
Oh, gotcha. It is hard to keep all these things straight...

emdupre commented:
It really, really is :sweat_smile: I'm happy to talk through this to make sure we're all on the same page before merging ! I'm hitting a memory error locally so I want to see if it replicates on circle, but I think this is close....

emdupre commented:
Ok, this is finally ready ! It's brought up a few subsequent issues I'd like to address (the occasional T2* coregistration seeming to have bad q/sform codes, updating the boilerplate, when to skullstrip the optimal combination), but overall this is a big step forward I think.

Thanks so much to  everyone, and in particular to @effigies for all of the time you've already taken to review / help this PR :hugs:

effigies commented on docs/workflows.rst:
```diff
@@ -592,3 +561,22 @@ A visualization of the AROMA component classification is also included in the HT
+and optimally weighted combination of all supplied single echo time series.
+This optimaly combined time series is then carried forward for all subsequent
+preprocessing steps.
+Optionally, if the ``--t2s_coreg`` flag is supplied, the T2* map is then used
```
`t2s-coreg` (hyphen, not underscore)

https://github.com/poldracklab/fmriprep/blob/8a4de9fccd28d06ce765df1a5aea4e4abb9ed0f2/fmriprep/cli/run.py#L117

effigies commented on fmriprep/__about__.py:
```diff
@@ -108,6 +108,8 @@
 
 LINKS_REQUIRES = [
+    'git+https://github.com/me-ica/tedana.git@'
+    'cb01d725d05f0102e2da89ad50c32905c68a79fd#egg=tedana',
```
Since we're merging to `multiecho` and not `master`, this isn't urgent, but it would be good to depend on a release version of tedana before we have fMRIPrep releases that depend on this commit.

emdupre replied:
:+1: yes, there's a growing list of things `tedana` needs, but cutting a release soon is definitely one of them

effigies commented on fmriprep/workflows/bold/base.py:
```diff
@@ -301,17 +298,15 @@ def init_func_preproc_wf(bold_file, ignore, freesurfer,
-        run_stc = False
+    # By default, force-bbr for t2s_coreg unless user specifies otherwise
+    if t2s_coreg and (use_bbr is not False):
+            use_bbr = True
```
Overindented.

effigies commented on fmriprep/workflows/bold/base.py:
```diff
@@ -301,17 +298,15 @@ def init_func_preproc_wf(bold_file, ignore, freesurfer,
-                       'multiecho BOLD data (not implemented).')
-        run_stc = False
+    # By default, force-bbr for t2s_coreg unless user specifies otherwise
+    if t2s_coreg and (use_bbr is not False):
```
How about `if t2s_coreg and use_bbr is None:`? If it's `True`, this branch has no effect anyway, and we can be more explicit that this is the "default" behavior. Also, can you add a note to the docstring (L97-99)?

effigies commented on fmriprep/workflows/bold/registration.py:
```diff
@@ -297,31 +299,47 @@ def init_bold_t1_trans_wf(freesurfer, mem_gb, omp_nthreads, use_fieldwarp=False,
-        (inputnode, bold_to_t1w_transform, [('bold_split', 'input_image')]),
         (gen_ref, bold_to_t1w_transform, [('out_file', 'reference_image')]),
+        # merge 3D volumes into 4D time series
+        (inputnode, merge, [('name_source', 'header_source')]),
```
Sorry, could you move this back above, just to reduce the diff to actually changed behavior?

emdupre replied:
It is changed behavior here, I think -- now it's called regardless of whether MEEPI data is provided, whereas before it was only called for non-MEEPI data

effigies replied:
Sorry, I just meant "above  the `(gen_ref, bold_to_t1w_transform, ` line". It's identical with the removed line 322.

emdupre replied:
ok, I can get rid of the identical comment ! sorry, I didn't realize I had duplicated that. or will removing that cause more headache for the diff ?

effigies replied:
Sorry, I don't understand the question. The latest update looks good, though.

effigies commented on fmriprep/workflows/bold/registration.py:
```diff
@@ -297,31 +299,47 @@ def init_bold_t1_trans_wf(freesurfer, mem_gb, omp_nthreads, use_fieldwarp=False,
-            (inputnode, merge_xforms, [('fieldwarp', 'in2')])
-        ])
+    # merge 3D volumes into 4D timeseries
+    merge = pe.Node(Merge(compress=use_compression), name='merge', mem_gb=mem_gb)
```
Could you move this back below `bold_to_t1w_transform`?

effigies commented:
Sorry, thought I'd already merged this...
